### ISSUE TITLE
TopNWordsInFile.java - Runtime Question

### ISSUE THREAD

**Issue by kxgoto1927** [jeffreyliguori/chi_tech_interviews#1]
Hey Jeff, I ran the TopNWordsInFile.java and the output between building a linked list as the value of the hashmap entry (Fast) and PriorityQueue(Slow) is inconsistent.  Sometimes Fast is slower and most of the time, even as input increases, they seem to be pretty close in value.

![image](https://user-images.githubusercontent.com/7820500/48008233-e25e6a00-e0de-11e8-9c21-8034b93df27a.png)

Ignore me if I'm reading the output wrong.  I'm under the assumption that lower the faster.

Further more, if i create a new method called `baseLineReadFile` which contain a single loop through the List, wouldn't that help us test whether our fast algo is within the upperbound?  Let me know if this is a bad test.

![image](https://user-images.githubusercontent.com/7820500/48008938-43d30880-e0e0-11e8-8c04-83c38b1f21a2.png)

In terms of space complexity, would our linked list solution have O(m) where m is the unique number of word count since we never delete the Node?

**Comment by jeffreyliguori** [comment]
Hey Kenta,

I didn't print that stuff in a particularly helpful way. The runtimes of
the two methods I have there are:

Slow: O(n + mlogk)
Fast: O(n + m), which is just O(n) since n >= m

where n is total number of words in the file, m is unique words in the
file, and k is the top k most frequent words you're looking for.

As such, I tested the runtimes with both m increasing and k increasing, to
see that both increase when you increased m but only the "slow" one
increased when you increased k. My output shows that as you increase m/n,
both get slower, but as you increase k (the --- with large n's ---
section), the slow version gets slower as k increases but fast stays mostly
the same.

The reason the "slow" version is usually faster is most likely because it
was implemented better; creating new objects is expensive, and my "fast"
method does a lot of that, whereas the priorityqueue just does a bunch of
array shuffling. However, as k increases, you'd expect slow to eventually
perform worse than fast, even though my old computer couldn't run inputs
big enough to prove it :)

Let me know if that's useful.
